components/brave_wallet/browser/account_discovery_manager.cc

```diff
@@ -0,0 +1,159 @@
+    const mojom::CoinType& coin_type,
+    const std::string& keyring_id,
+    const std::string& chain_id,
+    size_t discovery_account_index,
+    int attempts_left)
+    : coin_type(coin_type),
+      keyring_id(keyring_id),
+      chain_id(chain_id),
+      discovery_account_index(discovery_account_index),
+      attempts_left(attempts_left) {}
+
+AccountDiscoveryManager::DiscoveryContext::~DiscoveryContext() = default;
+
+AccountDiscoveryManager::AccountDiscoveryManager(
+    JsonRpcService* rpc_service,
+    KeyringService* keyring_service)
+    : json_rpc_service_(rpc_service), keyring_service_(keyring_service) {}
+
+void AccountDiscoveryManager::StartDiscovery() {
+  if (keyring_service_->IsKeyringCreated(mojom::kDefaultKeyringId)) {
+    AddDiscoveryAccount(std::make_unique<DiscoveryContext>(
+        mojom::CoinType::ETH, mojom::kDefaultKeyringId, mojom::kMainnetChainId,
+        keyring_service_->GetAccountsNumber(mojom::kDefaultKeyringId)
+            .value_or(0),
+        kDiscoveryAttempts));
+  }
+  if (IsFilecoinEnabled() &&
+      keyring_service_->IsKeyringCreated(mojom::kFilecoinKeyringId)) {
+    AddDiscoveryAccount(std::make_unique<DiscoveryContext>(
+        mojom::CoinType::FIL, mojom::kFilecoinKeyringId,
+        mojom::kFilecoinMainnet,
+        keyring_service_->GetAccountsNumber(mojom::kFilecoinKeyringId)
+            .value_or(0),
+        kDiscoveryAttempts));
+  }
+  if (IsSolanaEnabled() &&
+      keyring_service_->IsKeyringCreated(mojom::kSolanaKeyringId)) {
+    AddDiscoveryAccount(std::make_unique<DiscoveryContext>(
+        mojom::CoinType::SOL, mojom::kSolanaKeyringId, mojom::kSolanaMainnet,
+        keyring_service_->GetAccountsNumber(mojom::kSolanaKeyringId)
+            .value_or(0),
+        kDiscoveryAttempts));
+  }
+}
+
+AccountDiscoveryManager::~AccountDiscoveryManager() {}
+
+void AccountDiscoveryManager::AddDiscoveryAccount(
+    std::unique_ptr<DiscoveryContext> context) {
+  if (context->attempts_left <= 0) {
+    return;
+  }
+
+  auto addr = keyring_service_->GetDiscoveryAddress(
+      context->keyring_id, context->discovery_account_index);
+  if (!addr) {
+    NOTREACHED();
+    return;
+  }
+
+  auto chain_id = context->chain_id;
+  auto coin_type = context->coin_type;
+  if (context->coin_type == mojom::CoinType::ETH) {
+    json_rpc_service_->GetEthTransactionCount(
+        chain_id, addr.value(),
+        base::BindOnce(&AccountDiscoveryManager::OnEthGetTransactionCount,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(context)));
+  } else if (context->coin_type == mojom::CoinType::SOL) {
+    json_rpc_service_->GetSolanaBalance(
+        addr.value(), chain_id,
+        base::BindOnce(&AccountDiscoveryManager::OnResolveSolanaAccountBalance,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(context)));
+  } else if (context->coin_type == mojom::CoinType::FIL) {
+    json_rpc_service_->GetBalance(
```

Comment:
Checking balance would makes us miss the account that has transactions with zero balance.

Reply:
Proper method to use is limited https://github.com/filecoin-project/lotus/issues/9728

Reply:
ok, please add a comment